src/chunks.rs

```diff
@@ -0,0 +1,191 @@
+use anyhow::Context;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Cursor, Seek, SeekFrom};
+use std::path::Path;
+
+fn split(chunk: u64, size: u64) -> Vec<u64> {
+    let e = size / chunk + 1.min(size % chunk);
+    (0..e).map(|i| i * chunk).collect()
+}
+
+#[derive(Debug)]
+pub struct Chunk<C> {
+    lines: std::io::Lines<C>,
+    pos: u64,
+    end: u64,
+}
+
+impl<C> Chunk<C> {
+    pub fn new(
+        mut c: C,
+        chunk: usize,
+        mut current: u64,
+        start: u64,
+        size: u64,
+    ) -> anyhow::Result<(Self, u64)>
+    where
+        C: Seek + BufRead,
+    {
+        let skip = if current > start {
+            true
+        } else if start != current {
+            c.seek(SeekFrom::Start(start - 1))?;
+            let mut buf = [0 as u8; 1];
+            if let Ok(1) = c.read(&mut buf) {
+                buf[0] != b'\n'
+            } else {
+                false
+            }
+        } else {
+            false
+        };
+
+        c.seek(SeekFrom::Start(start))?;
+        current = if skip {
+            let mut skip_leader = String::new();
+            let _ = c.read_line(&mut skip_leader)?;
+            start + skip_leader.len() as u64
+        } else {
+            start
+        };
+        let lines = c.lines();
+        let c = Self {
+            lines,
+            pos: current,
+            end: size.min(start + chunk as u64),
+        };
+        Ok((c, current))
+    }
+}
+
+impl<T> Iterator for Chunk<T>
+where
+    T: Seek + BufRead,
+{
+    type Item = String;
+    fn next(&mut self) -> Option<String> {
+        if self.pos >= self.end {
+            return None;
+        }
+        match self.lines.next() {
+            Some(Ok(l)) => {
+                self.pos += l.len() as u64 + 1;
+                Some(l)
+            }
+            _ => None,
+        }
+    }
+}
+
+pub struct Chunks<T> {
+    files: Box<dyn Iterator<Item = anyhow::Result<Cursor<T>>> + Send>,
+    current: u64,
+    starts: std::vec::IntoIter<u64>,
+    chunk: usize,
+    size: u64,
+}
+
+pub fn chunks(path: &Path, chunk: usize, size: u64) -> Chunks<memmap::Mmap> {
+    let path = path.to_path_buf();
+    let it = (0..usize::MAX).map(move |_| {
```

Comment:
I don't understand what the purpose of this loop is. In particular, why are we iterating over the `0..usize::MAX` range in the first place? Also why do we need this? If we've gotten the slice out of the memory map we can easily chunk that up.

Reply:
We split the file in chunks. For this we open the file and seek to the start position. The iterator is a way to open unlimited number of files for the path. The asumption is that we have at most (in the order of) number of cpus files open at any given time as rayon is handing out work.

Reply:
Why don't we just open the `File` as part of the instantiation of a `Chunk`?

Reply:
You need a path to open a file. In the tests I want to run in mem without file io, so I need some why to avoid refering to paths. There's probably a better way to do this.